Userland/Libraries/LibAudio/ConnectionFromClient.h

```diff
@@ -1,40 +1,102 @@
 /*
  * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2022, kleines Filmröllchen <filmroellchen@serenityos.org>
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
 #pragma once
 
+#include <AK/Concepts.h>
+#include <AK/FixedArray.h>
+#include <AK/NonnullOwnPtr.h>
+#include <AK/OwnPtr.h>
 #include <AudioServer/AudioClientEndpoint.h>
 #include <AudioServer/AudioServerEndpoint.h>
+#include <LibAudio/Buffer.h>
+#include <LibAudio/UserSampleQueue.h>
+#include <LibCore/EventLoop.h>
+#include <LibCore/Object.h>
 #include <LibIPC/ConnectionToServer.h>
+#include <LibThreading/Thread.h>
 
 namespace Audio {
 
-class LegacyBuffer;
-
 class ConnectionFromClient final
     : public IPC::ConnectionToServer<AudioClientEndpoint, AudioServerEndpoint>
     , public AudioClientEndpoint {
     IPC_CLIENT_CONNECTION(ConnectionFromClient, "/tmp/portal/audio")
 public:
-    void enqueue(LegacyBuffer const&);
-    bool try_enqueue(LegacyBuffer const&);
-    void async_enqueue(LegacyBuffer const&);
+    virtual ~ConnectionFromClient() override;
+
+    // Both of these APIs are for convenience and when you don't care about real-time behavior.
+    // They will not work properly in conjunction with realtime_enqueue.
+    // If you don't refill the buffer in time with this API, the last shared buffer write is zero-padded to play all of the samples.
+    template<ArrayLike<Sample> Samples>
+    ErrorOr<void> async_enqueue(Samples&& samples)
+    {
+        return async_enqueue(TRY(FixedArray<Sample>::try_create(samples.span())));
+    }
+
+    ErrorOr<void> async_enqueue(FixedArray<Sample>&& samples);
+
+    // Like async_enqueue except it yields the thread until all samples have played.
+    template<ArrayLike<Sample> Samples>
+    ErrorOr<void> enqueue(Samples& samples)
+    {
+        TRY(async_enqueue(samples));
+
+        while (remaining_samples() > 0)
+            // FIXME: better solution needed
+            sched_yield();
+
+        return {};
+    }
```

Comment:
this is unused anyways, sync enqueue is a bad idea and only caused issues in Piano & others. ill remove it